icu4c/source/i18n/gregocal.cpp

```diff
@@ -332,7 +332,7 @@ GregorianCalendar::setGregorianChange(UDate date, UErrorCode& status)
     // normalized cutover is in pure date milliseconds; it contains no time
     // of day or timezone component, and it used to compare against other
     // pure date values.
-    int32_t cutoverDay = (int32_t)ClockMath::floorDivide(fGregorianCutover, (double)kOneDay);
+    int64_t cutoverDay = ClockMath::floorDivide(fGregorianCutover, (double)kOneDay);
```

Comment:
I am not sure if your change fixes the issue completely.

This is what Fredrik found last year:
> The float-cast-overflow UBSan [undefined-behavior sanitizer] check ...
>
> ... setGregorianChange() is written to expect numeric overflow for certain inputs, in which case UBSan will erroneously fail like this:
>
> .../icu/source/i18n/gregocal.cpp:335:35: runtime error: -2.08066e+300 is outside the range of representable values of type 'int'

By casting the double result to int64_t rather than int32_t, you widen the range of values that fit, but doubles can still be much larger. The code (and the comment below) seems to assume that the UDate can be no lower than "Long.MIN_VALUE" which would be true in Java where we do use long, but not in C++ where we use UDate=double, unless the cutover setter limits the range of values.

Also, the code following the floorDivide() still has a hack of detecting an overflow by checking if dividing a negative value yields a positive quotient, but it seems like there is a 50/50 chance that an overflow would give us a negative value (assuming it just stuffs the double bits into the int64_t).

Wouldn't it be better to put the quotient (cutoverDay) into a double, do a range check in "double" space to detect what we need (I am not quite sure what the proper replacement for the "Handle the rare case..." hack should be), and prevent an overflow later when converting to the int32_t fCutoverJulianDay?

Reply:
Re-reading the code... This is the setGregorianChange() public API, so there is nothing else to limit the range of values.

And right after the floorDivide we re-multiply to get the fNormalizedGregorianCutover to be the start of the day corresponding to the caller's given point in time.

So what this apparently tried to do is to deal with where date-as-a-double divided by millis-per-day does not fit into a long, and compensate.

How about we change the cutoverDay into a double, and if it's below INT32_MIN or above INT32_MAX we adjust both the fGregorianCutover and the fNormalizedGregorianCutover to limit * kOneDay -- and remove the rest of the overflow detection stuff?

Maybe
```
double cutoverDay = ClockMath::floorDivide(fGregorianCutover, (double)kOneDay);
if (cutoverDay < INT32_MIN) {
    fGregorianCutover = fNormalizedGregorianCutover = (double)INT32_MIN * kOneDay;
} else if (cutoverDay > INT32_MAX) {
    fGregorianCutover = fNormalizedGregorianCutover = (double)INT32_MAX * kOneDay;
} else {
    fNormalizedGregorianCutover = cutoverDay * kOneDay;
}
```

Reply:
there's also a floorDivide that returns an int64_t

Reply:
Yes, but signature matching is on the parameters, not on where the return value ends up.

Given that the inputs are both doubles, calling the double version looks proper, and we should deal with the double result.

Reply:
OK. Working on a fix. Also seems we can just do `floorDivide(date, …)`